python/ray/autoscaler/_private/aws/config.py

```diff
@@ -959,6 +963,78 @@ def _configure_node_cfg_from_launch_template(
     return node_cfg
 
 
+def _configure_cluster_event_notifications(
+        config: Dict[str, Any]) -> Dict[str, Any]:
+    """If `cluster_notifications` is provided, an SNS callback handler will be added to
+    cluster creation events. The callback handler will send the event ID and publish
+    a notification to the SNS topic specified at `cluster_notifications.sns_topic_arn`.
+    The SNS topic must be in the same AWS account associated to the autoscaler cluster role.
+
+    Message parameters can be specified in `cluster_notifications.parameters`, which
+    will be included into the payload of each SNS notification.
+
+    Args:
+        config (Dict[str, Any]): config to bootstrap
+
+    Returns:
+        config (Dict[str, Any]): The input config with cluster event notification
+        data merged into the node config of all available node types. If no
+        cluster event notification data is found, then the config is returned
+        unchanged.
+    """
+    # create a copy of the input config to modify
+    config = copy.deepcopy(config)
+    provider = config['provider']
+
+    # TODO: Need a way to generalize this so that users can plug in their own custom callback logic
+    #  or have a dynamic config
+
+    if provider and provider.get('cluster_notifications'):
+        cluster_notifications_map = provider['cluster_notifications']
+        parameters = cluster_notifications_map['parameters']
+        sns_topic_arn = cluster_notifications_map['sns_topic_arn']
+        callback_kwargs = {
+            "sns_topic_arn": sns_topic_arn,
+            "params": parameters,
+        }
+        for i, event_name in enumerate(
+                CreateClusterEvent.__members__.values()):
+            global_event_system.add_callback_handler(
+                event_name, _cluster_event_notifications_callback, snsu,
+                **callback_kwargs)
+    return config
+
+
+def _cluster_event_notifications_callback(
+        sns_client, event_data: Dict[str, Any], **kwargs):
+    """Generic callback for sending Ray cluster setup event data to an SNS topic.
+
+    Args:
+        sns_client: Amazon SNS client for publishing to an SNS topic
+        event_data: Ray cluster setup event data. This contains the event name, enum ID, and
+            may also contain additional metadata (i.e. the initialization or setup command used
+            during this setup step)
+        **kwargs: Keyword arguments that were injected into `_EventSystem.add_callback_handler`
+    """
+    try:
+        # create a copy of the event data to modify
+        event_dict = copy.deepcopy(event_data)
+        event: CreateClusterEvent = event_dict.pop('event_name')
+        sns_topic_arn, params = kwargs['sns_topic_arn'], kwargs['params']
+        message = {
+            **params,
+            "setupEventMetadata": event_dict,
+            "stateSequence": event.value - 1,  # zero-index sequencing
+            "timestamp": round(time.time() * 1000)
```

Comment:
We could add additional two fields here to distinguish between head node and worker node: is_head_node: boolean, node_id: str.
Quickly checked and the following events are sent by every node in [updater.py](https://github.com/amzn/amazon-ray/blob/main/python/ray/autoscaler/_private/updater.py):
`ssh_control_acquired`
`run_initialization_cmd`
`run_setup_cmd`
`start_ray_runtime`
`start_ray_runtime_completed`
We can use the self.is_head_node and self.node_id in updater.py when executing callback as part of event_data.

Reply:
Good to know, thanks Miranda! 

On another note, what is the preferred process for testing these changes specifically on worker nodes? Testing it on the head node is simple enough, since the local Ray installation has these changes, but the worker nodes will still be running off an unchanged Ray environment. It sounds like I can manually create an AMI with these callback changes, and have the worker nodes use them, but maybe there are easier ways to test this?

Reply:
Yeah, you can mount your local dev folder in file_mounts. Then replace this folder under where ray is installed on every node. Check out this [example](https://code.amazon.com/packages/SunGate/blobs/mainline/--/configuration/autoscaler/fast-deployment-of-python-codes.yaml).